certbot/hooks.py

```diff
@@ -57,30 +57,70 @@ def validate_hook(shell_cmd, hook_name):
 
 
 def pre_hook(config):
-    "Run pre-hook if it's defined and hasn't been run."
+    """Run pre-hooks if they exist and haven't already been run.
+
+    When Certbot is running with the renew subcommand, this function
+    runs any hooks found in the config.renewal_pre_hooks_dir (if they
+    have not already been run) followed by any pre-hook in the config.
+    If hooks in config.renewal_pre_hooks_dir are run and the pre-hook in
+    the config is a path to one of these scripts, it is not run twice.
+
+    :param configuration.NamespaceConfig config: Certbot settings
+
+    """
+    if config.verb == "renew":
+        for hook in list_hooks(config.renewal_pre_hooks_dir):
+            _run_pre_hook_if_necessary(hook)
+
     cmd = config.pre_hook
-    if cmd and cmd not in pre_hook.already:
-        logger.info("Running pre-hook command: %s", cmd)
-        _run_hook(cmd)
-        pre_hook.already.add(cmd)
-    elif cmd:
-        logger.info("Pre-hook command already run, skipping: %s", cmd)
+    if cmd:
+        _run_pre_hook_if_necessary(cmd)
 
 pre_hook.already = set()  # type: ignore
 
 
+def _run_pre_hook_if_necessary(command):
+    """Run the specified pre-hook if we haven't already.
+
+    If we've already run this exact command before, a message is logged
+    saying the pre-hook was skipped.
+
+    :param str command: pre-hook to be run
+
+    """
+    if command in pre_hook.already:
```

Comment:
I think we should use `os.path.realpath(command)` here to prevent symlinked scripts being run twice.

Reply:
Initially I agreed with you, +1'd these comments, and starting working on this, but this change could theoretically break things for people if they're relying on the name of `argv[0]` in their script right? Doing this is unusual but certainly not unheard of. Do we want to risk this?

Reply:
I guess we'll have to choose one of the possible corner cases :) However I feel like relying on `argv[0]` would be a bad practice.

Reply:
Perhaps it is, but scripts such as `apache2ctl` (at least on Debian systems) and our own `certbot-auto` do this.

This would also be a change in behavior for people who aren't even using our hook directories. I agree it's a trade-off, but since we already have made the decision one way, I think we should stick with it rather than changing behavior now.

Reply:
The scripts themselves use and rely on `argv[0]` and that's fine, but I would think that a script like that would get confused explicitly in the case where it would be run through a symlink instead of the actual script path, something that this modification would prevent. 

However there most likely is something that I'm missing, and I trust on your reasoning on this subject, so if you are confident that this is the right way forward we can leave it as it is.

Reply:
Yes, I'd prefer to leave this as is.

Unless we use `realpath` only for directory hooks (which adds unnecessary edge cases and complexity to our hooks system), this is a change in behavior for config hooks which I'd like to avoid making. It's unlikely to matter, but unless we have a good reason to suddenly change the behavior for people, I'd prefer we didn't do this. Symlinks are often the way I've seen differences based on `argv[0]` implemented such as `apache2ctl` on Debian/Ubuntu. It seems extremely unlikely anyone's hook scripts are that complex, but who knows.